eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggererSpec.scala

```diff
@@ -0,0 +1,167 @@
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import com.typesafe.config.ConfigFactory
+import fr.acinq.bitcoin.scalacompat.ByteVector32
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.blockchain.CurrentBlockHeight
+import fr.acinq.eclair.channel.{CMD_GET_CHANNEL_STATE, NEGOTIATING, RES_GET_CHANNEL_STATE}
+import fr.acinq.eclair.{BlockHeight, TestConstants, randomKey}
+import fr.acinq.eclair.payment.relay.AsyncPaymentTriggerer._
+import fr.acinq.eclair.io.{Peer, PeerConnected, Switchboard}
+import fr.acinq.eclair.io.Switchboard.GetPeerInfo
+import org.scalatest.Outcome
+import org.scalatest.funsuite.FixtureAnyFunSuiteLike
+
+import scala.concurrent.duration.DurationInt
+
+class AsyncPaymentTriggererSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("application")) with FixtureAnyFunSuiteLike {
+
+  case class FixtureParam(remoteNodeId: PublicKey, switchboard: TestProbe[Switchboard.GetPeerInfo], channelProbes: Seq[TestProbe[CMD_GET_CHANNEL_STATE]], probe: TestProbe[Result], triggerer: ActorRef[Command]) {
+    def channels: Set[akka.actor.ActorRef] = channelProbes.map(_.ref.toClassic).toSet
+  }
+
+  override def withFixture(test: OneArgTest): Outcome = {
+    val remoteNodeId = TestConstants.Alice.nodeParams.nodeId
+    val switchboard = TestProbe[Switchboard.GetPeerInfo]("switchboard")
+    val channelProbes = Seq(TestProbe[CMD_GET_CHANNEL_STATE]("channel1"), TestProbe[CMD_GET_CHANNEL_STATE]("channel2"))
+    val probe = TestProbe[Result]()
+    val triggerer = testKit.spawn(AsyncPaymentTriggerer())
+    triggerer ! Start(switchboard.ref)
+    withFixture(test.toNoArgTest(FixtureParam(remoteNodeId, switchboard, channelProbes, probe, triggerer)))
+  }
+
+  test("remote node does not connect before timeout") { f =>
+    import f._
+
+    triggerer ! Watch(probe.ref, remoteNodeId, paymentHash = ByteVector32.Zeroes, timeout = BlockHeight(100))
+    assert(switchboard.expectMessageType[GetPeerInfo].remoteNodeId == remoteNodeId)
+
+    // We haven't reached the timeout yet.
+    system.eventStream ! EventStream.Publish(CurrentBlockHeight(BlockHeight(99)))
+    probe.expectNoMessage(100 millis)
+
+    // We exceed the timeout (we've missed blocks).
+    system.eventStream ! EventStream.Publish(CurrentBlockHeight(BlockHeight(110)))
+    probe.expectMessage(AsyncPaymentTimeout)
+
+    // Only get the timeout message once.
+    system.eventStream ! EventStream.Publish(CurrentBlockHeight(BlockHeight(111)))
+    probe.expectNoMessage()
+  }
+
+  test("duplicate watches should emit only one trigger") { f =>
+    import f._
+
+    // create two identical watches
+    triggerer ! Watch(probe.ref, remoteNodeId, paymentHash = ByteVector32.Zeroes, timeout = BlockHeight(100))
+    assert(switchboard.expectMessageType[GetPeerInfo].remoteNodeId == remoteNodeId)
+    triggerer ! Watch(probe.ref, remoteNodeId, paymentHash = ByteVector32.Zeroes, timeout = BlockHeight(100))
+    switchboard.expectNoMessage()
```

Comment:
It isn't really useful to test the behavior of the underlying `PeerReadyNotifier`, it is already covered by the `PeerReadyNotifierSpec`. Calls to `expectNoMessage()` can be expensive when we make too many of them, because they slow down the test suite (which we fight hard to keep below 20 minutes on CI). I removed some of these redundants checks in https://github.com/ACINQ/eclair/commit/ac5118b0b5dc1c1f543bca076a43aba5f4ea9ef3